### PR TITLE
fix build omitempty and impossible pointers

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/3212

Build validation rules didn't match the types.  It was possible to build invalid objects  in ways that didn't make sense.  If a small, unshared pointer can never be nil, it shouldn't be a pointer (`from`).  Also fixed json tags so that unconditionally required fields can no longer be omitted.

@bparees

### CODE REVIEW

File: pkg/build/api/v1/types.go [hunk 1]
```diff
@@ -212,7 +212,7 @@ const (
 type CustomBuildStrategy struct {
 	// From is reference to an ImageStream, ImageStreamTag, or ImageStreamImage from which
 	// the docker image should be pulled
-	From *kapi.ObjectReference `json:"from,omitempty" description:"reference to an image stream, image stream tag, or image stream image from which the Docker image should be pulled"`
+	From kapi.ObjectReference `json:"from" description:"reference to an image stream, image stream tag, or image stream image from which the Docker image should be pulled"`
```
Comment (liggitt) [inline pkg/build/api/v1/types.go]: careful, this would have accepted `null` before

File: pkg/build/api/v1/types.go [hunk 2]
```diff
@@ -249,7 +249,7 @@ type DockerBuildStrategy struct {
 type SourceBuildStrategy struct {
 	// From is reference to an ImageStream, ImageStreamTag, or ImageStreamImage from which
 	// the docker image should be pulled
-	From *kapi.ObjectReference `json:"from,omitempty" description:"reference to an image stream, image stream tag, or image stream image from which the Docker image should be pulled"`
+	From kapi.ObjectReference `json:"from" description:"reference to an image stream, image stream tag, or image stream image from which the Docker image should be pulled"`
```
Comment (liggitt) [inline pkg/build/api/v1/types.go]: same here

Comment (deads2k) [inline pkg/build/api/v1/types.go]: > same here

joy, so conversions?  grrr....

Comment (deads2k) [inline pkg/build/api/v1/types.go]: hmmm.. .accepted, but it would have been uncreatable (validation would have rejected).  I'm going to claim this is ok.

Comment (liggitt) [inline pkg/build/api/v1/types.go]: interesting... that seems ok... @bparees @smarterclayton concur?

File: pkg/build/api/validation/validation.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ func ValidateBuild(build *buildapi.Build) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
 	allErrs = append(allErrs, validation.ValidateObjectMeta(&build.ObjectMeta, true, validation.NameIsDNSSubdomain).Prefix("metadata")...)
 
-	allErrs = append(allErrs, validateBuildParameters(&build.Parameters).Prefix("parameters")...)
+	allErrs = append(allErrs, validateBuildParameters(&build.Parameters).Prefix("spec")...)
```
Comment (bparees) [inline pkg/build/api/validation/validation.go]: why?  the field name is parameters, not spec.

File: pkg/build/api/validation/validation.go [hunk 4]
```diff
@@ -51,8 +51,8 @@ func ValidateBuildConfig(config *buildapi.BuildConfig) fielderrors.ValidationErr
 			}
 		}
 	}
-	allErrs = append(allErrs, validateBuildParameters(&config.Parameters).Prefix("parameters")...)
-	allErrs = append(allErrs, validateBuildConfigOutput(&config.Parameters.Output).Prefix("parameters.output")...)
+	allErrs = append(allErrs, validateBuildParameters(&config.Parameters).Prefix("spec")...)
+	allErrs = append(allErrs, validateBuildConfigOutput(&config.Parameters.Output).Prefix("spec.output")...)
```
Comment (bparees) [inline pkg/build/api/validation/validation.go]: same q.

Comment (liggitt) [inline pkg/build/api/validation/validation.go]: this is what surfaces to API users to tell them what's wrong with what they submitted. it is supposed to match the serialized field name

File: pkg/build/api/validation/validation.go (hunk 3 above)
Comment (bparees) [inline pkg/build/api/validation/validation.go]: nm, forgot that v1 is ahead of internal on these objects.

File: pkg/build/api/v1/types.go (hunk 2 above)
Comment (bparees) [inline pkg/build/api/v1/types.go]: yeah i was going to make the same claim (that validation would/should have rejected it anyway)

**Comment by bparees** [comment]
one thing that bugs me about this is that now From is sometimes a pointer and sometimes not.  That strikes me as somewhat hideous.  Otherwise I guess it's ok.

**Comment by deads2k** [comment]
> one thing that bugs me about this is that now From is sometimes a pointer and sometimes not. That strikes me as somewhat hideous.

It bothers me a bit, but I'd rather have the correct types so that the compiler can protect me from creating some kinds of invalid objects.  I console myself by saying that different things are different.

**Comment by deads2k** [comment]
bump

**Comment by bparees** [comment]
[test]

**Comment by deads2k** [comment]
docker registry flake.  re[test]

**Comment by deads2k** [comment]
rebased.  Is the only concern here that we get a clean test run?

**Comment by bparees** [comment]
nah this is fine.  [merge]

**Comment by danmcp** [comment]
@deads2k Needs a rebase

**Comment by deads2k** [comment]
rebased [merge]

**Comment by deads2k** [comment]
rebased [merge]

**Comment by deads2k** [comment]
hack/test-cmd.sh works locally,  [merge]

**Pull request merged**